lib/parser/prescan.cc

```diff
@@ -282,26 +282,38 @@ void Prescanner::SkipToEndOfLine() {
   }
 }
 
+bool Prescanner::MustSkipToEndOfLine() const {
+  if (inFixedForm_ && column_ > fixedFormColumnLimit_ && !tabInCurrentLine_) {
+    return true;  // skip over ignored columns in right margin (73:80)
+  } else if (*at_ == '!' && !inCharLiteral_) {
+    return true;  // inline comment goes to end of source line
+  } else {
+    return false;
+  }
+}
+
 void Prescanner::NextChar() {
   CHECK(*at_ != '\n');
   ++at_, ++column_;
   if (inPreprocessorDirective_) {
     SkipCComments();
   } else {
-    bool rightMarginClip{
-        inFixedForm_ && column_ > fixedFormColumnLimit_ && !tabInCurrentLine_};
-    bool skipping{rightMarginClip || (*at_ == '!' && !inCharLiteral_)};
-    if (skipping) {
+    bool mightNeedSpace{false};
+    if (MustSkipToEndOfLine()) {
       SkipToEndOfLine();
+    } else {
+      mightNeedSpace = *at_ == '\n';
     }
     while (*at_ == '\n' || *at_ == '&') {
-      bool mightNeedSpace{*at_ == '\n' && !skipping};
-      if (inFixedForm_) {
-        if (!FixedFormContinuation(mightNeedSpace)) {
-          return;
+      bool continued{inFixedForm_ ? FixedFormContinuation(mightNeedSpace)
+                                  : FreeFormContinuation()};
+      if (continued) {
+        if (MustSkipToEndOfLine()) {
+          SkipToEndOfLine();
         }
-      } else if (!FreeFormContinuation()) {
-        return;
+        mightNeedSpace = false;
+      } else {
+        break;
       }
```

Comment:
The logic *might* be clearer like this:
```
if (!continued) {
   break;
}
if (MustSkipToEndOfLine()) { ...
```

Reply:
Please take another look and see whether clarity has been improved for you.

Reply:
> Please take another look and see whether clarity has been improved for you.

Yes, thank you.